contributors/design-proposals/storage/csi-migration.md

```diff
@@ -233,46 +278,112 @@ and how to redeploy a CSI Driver in a cluster.
+	VolumePluginMechanism VolumePluginMechanism `json:"volumePluginMechanism"`
+}
+```
+
+The important fields to note are `Available` which allows the Attach/Detach
+controller to tell whether the driver is available on each node or not. As well
+as `VolumePluginMechanism` which is set by the Kubelet depending on the feature
+flags on the Kubelet for CSI Migration which tell the Attach/Detach controller
+whether it should use the fallback in-tree code or whether it should go through
+the csi-plugin migration path.
+
+### Feature Flags
+
+ADC and Kubelet use the "same" feature flags, but in reality they are passed in
+to each binary seperately. There will be a feature flag per driver as well as
+one for CSIMigration in general.
+
+Kubelet will use its own feature flags to determine whether to use the in-tree
+or csi backend for Kubelet storage lifecycle operations, as well as to populate
+the `VolumePluginMechanism` field of `CSINodeInfo` for the driver for the node
+that Kubelet is running on.
+
+The ADC will also use its own feature flags to help make the determination
+whether to use in-tree or CSI backend for ADC storage lifecycle operations. The
+other component to help determine which backend to use will be outlined below in
+the "ADC and Kubelet CSI/In-tree Sync" section.
+
+### ADC and Kubelet CSI/In-tree Sync
+
+Some plugins have subtly different behavior on both ADC and Kubelet side between
+in-tree and CSI implementations. Therefore it is important that if the ADC is to
+use the in-tree implementation, the Kubelet must as well - and if the ADC is to
+use the CSI Migrated implementation, the Kubelet must as well. Therefore we will
+implement a mechanism to keep the ADC and the Kubelet in sync about the Kubelets
+abilities as well as the feature gates active in each.
+
+In order for the ADC controller to have the requisite information from the
+Kubelet to make informed decisions the Kubelet must propagate the
+`VolumePluginMechanism` information for each potentially migrated driver on
+Kubelet startup and be considered `NotReady` until that information is synced to
+the API server. This gives is the following guarantees:
+* If `CSINodeInfo` for a particular driver does not exist, then ADC can infer
+  the Kubelet is not at a version with migration logic and should therefore
+  fall-back to in-tree implementation
+* If `CSINodeInfo` exists, and `VolumePluginMechanism` is `in-tree`, then ADC
+  can infer Kubelet has migration logic however the Feature Flag for that
+  particular driver is `off` and the ADC should therefore fall-back to in-tree
+  storage implementation
+* If `CSINodeInfo` exists, and `VolumePluginMechanism` is `csi-plugin`, then ADC
+  can infer Kubelet has migration logic and the Feature Flag for that particular
+  driver is `on` and the ADC should therefore use the csi-plugin migration
+  implementation
+* If `CSINodeInfo` exists, and `VolumePluginMechanism` is `csi-plugin` but the
```

Comment:
Kubelet has no way of inferring state of the controller. The most important direction is ADC inferring state of Kubelet since ADC can be up to 2 versions higher than Kubelet version. ADC could have Migration on by default while Kubelet binary still has no concept of Migration whatsoever